File: backend/src/main/java/tn/steg/backend/deliverables/entity/Deliverable.java
package tn.steg.backend.deliverables.entity;

import jakarta.persistence.*;
import lombok.*;
import tn.steg.backend.candidates.entity.Candidate;
import tn.steg.backend.departments.entity.Supervisor;
import tn.steg.backend.internships.entity.Internship;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "deliverables")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Deliverable {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private java.util.UUID id;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private DeliverableStatus status = DeliverableStatus.DRAFT;

    @Column(name = "storage_key")
    private String storageKey;

    private String bucket;

    @Column(name = "object_id")
    private String objectId;

    private String checksum;

    @Column(name = "mime_type")
    private String mimeType;

    private Long size;

    @Builder.Default
    private Integer version = 1;

    @Column(name = "submitted_date")
    private LocalDate submittedDate;

    @Column(name = "validated_at")
    private LocalDateTime validatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "internship_id", nullable = false)
    private Internship internship;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "submitted_by_id")
    private Candidate submittedBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "validated_by_id")
    private Supervisor validatedBy;
}

File: mobile/lib/core/models/deliverable.dart
/// Backend `DeliverableResponse` from `/deliverables/internship/{id}`.
class Deliverable {
  const Deliverable({
    required this.id,
    required this.title,
    this.description,
    required this.status,
    this.storageKey,
    this.mimeType,
    this.size,
    this.version,
    this.submittedDate,
    this.internshipId,
    this.submittedById,
    this.submittedByName,
  });

  final String id;
  final String title;
  final String? description;
  final String status;
  final String? storageKey;
  final String? mimeType;
  final int? size;
  final int? version;
  final DateTime? submittedDate;
  final String? internshipId;
  final String? submittedById;
  final String? submittedByName;

  factory Deliverable.fromJson(Map<String, dynamic> json) => Deliverable(
        id: json['id'] as String,
        title: json['title'] as String? ?? '',
        description: json['description'] as String?,
        status: json['status'] as String? ?? '',
        storageKey: json['storageKey'] as String?,
        mimeType: json['mimeType'] as String?,
        size: (json['size'] as num?)?.toInt(),
        version: (json['version'] as num?)?.toInt(),
        submittedDate: json['submittedDate'] != null
            ? DateTime.parse(json['submittedDate'] as String)
            : null,
        internshipId: json['internshipId'] as String?,
        submittedById: json['submittedById'] as String?,
        submittedByName: json['submittedByName'] as String?,
      );

  bool get isSubmitted => status.toUpperCase() == 'SUBMITTED';
  bool get isValidated => status.toUpperCase() == 'VALIDATED';
}
